packages/compass-import-export/src/components/export-select-fields.tsx

```diff
@@ -0,0 +1,261 @@
+      newFields[`${evt.target.name}`] ^= 1; // Flip 1/0 to its opposite.
+      updateSelectedFields(newFields);
+    },
+    [updateSelectedFields, fields]
+  );
+
+  const handleHeaderCheckboxChange = useCallback(() => {
+    const newFields = Object.assign({}, fields);
+
+    if (isEveryFieldChecked) {
+      Object.keys(newFields).map((f) => (newFields[f] = 0));
+    } else {
+      Object.keys(newFields).map((f) => (newFields[f] = 1));
+    }
+
+    updateSelectedFields(newFields);
+  }, [isEveryFieldChecked, updateSelectedFields, fields]);
+
+  const handleAddFieldSubmit = useCallback(
+    (evt) => {
+      if (evt.key === 'Enter') {
+        const obj = {
+          [evt.target.value]: 1,
+        };
+        // Assign current entry to the end of the fields list.
+        const newFields = Object.assign({}, fields, obj);
+
+        updateSelectedFields(newFields);
+      }
+    },
+    [updateSelectedFields, fields]
+  );
+
+  useEffect(() => {
+    if (!autoScrollNewFieldInput) {
+      return;
+    }
+
+    if (newFieldRef.current) {
+      // Focus and scroll to the add new field input.
+      newFieldRef.current.scrollIntoView();
+      newFieldRef.current.focus();
+    }
+
+    setAutoScrollNewFieldInput(false);
+  }, [autoScrollNewFieldInput]);
+
+  useEffect(() => {
+    if (
+      lastRenderedFieldsLength.current !== 0 &&
+      fieldKeys.length > lastRenderedFieldsLength.current
+    ) {
+      setAutoScrollNewFieldInput(true);
+    }
+
+    lastRenderedFieldsLength.current = fieldKeys.length;
+  }, [fieldKeys]);
+
+  return (
+    <>
+      <div className={headerContainerStyles}>
+        <Body weight="medium">Select Fields</Body>
+        <Body>
+          The fields displayed are from a sample of documents in the collection.
+          To ensure all fields are exported, add missing field names.
+        </Body>
+        <div>
+          <Button
+            variant="primary"
+            leftGlyph={<Icon glyph="Plus" />}
+            size="xsmall"
+            onClick={onAddNewFieldButtonClicked}
+          >
+            Add new field
+          </Button>
+        </div>
+      </div>
+
+      <div className={tableContainerStyles}>
+        <Table
+          data={fieldKeys.map((field, index) => ({
+            checked: !!fields[field],
+            field,
+            index,
+          }))}
+          columns={[
+            <TableHeader
+              className={smallCellContainerStyle}
+              key="checkbox"
+              label={
+                <Checkbox
+                  aria-label={
+                    isEveryFieldChecked
+                      ? 'Deselect all fields'
+                      : 'Select all fields'
+                  }
+                  title={
+                    isEveryFieldChecked
+                      ? 'Deselect all fields'
+                      : 'Select all fields'
+                  }
+                  checked={isEveryFieldChecked}
```

Comment:
I think you might want to set `indeterminate` here if a subset of the fields are checked?

https://github.com/mongodb-js/compass/blob/7984c2abefd94e1d215275be2ae07b6b53338a2d/packages/compass-connection-import-export/src/components/select-table.tsx#L67

(Fwiw, that SelectTable component was written with the intent to be somewhat generically usable if we decide so, not tied to connection import/export. I’m not sure how easy it is to actually re-use it, though, esp. with the logic for adding additional fields here.)

Reply:
Good idea, going to look into it more in the upcoming import/export work. Having the add new field column is a bit of a hack of the table.